### DISCUSSION TITLE
[Proposal] Safer synchronization primitive syntax

### DISCUSSION THREAD

**Discussion by pzaj2** [dotnet/csharplang#6466] [Language Ideas]
I'd like to propose a new piece of syntax to C# that would improve the safety of consuming `SemaphoreSlim` and any other synchronization primitive or introduce a new synchronization primitive that would be safer to use.

### Premise

Currently, what you will sometimes see is code similar to:

```csharp
public class SynchronizedAuthorizationClient {
    private static readonly SemaphoreSlim _semaphore = new(1,1);
    ...

    public async Task<string?> GetAccessToken()
    {
         try {
             await _semaphore.WaitAsync();
         
             var cachedToken = _tokenStorage.GetToken();

            if (cachedToken != null) {
                return cachedToken;
            }

            var token = await _identityProviderClient.GetToken();
            _tokenStorage.SetToken(token);  

            return token;
        } finally {
             _semaphore.Release();
         }
    }
}
```

The code above is pretty simple but on many occasions, it will become more complicated than that. That will also make releasing the semaphore harder to manage in some situations and sometimes result in not releasing the semaphore, leading to potential deadlocks. This is especially hard to debug when the semaphore is used across multiple classes that need to be synchronized.

In the above scenario (and many others I came across in the past), `lock` cannot be applied due to asynchronous operations being performed inside the synchronized code block. I would, therefore, like to propose a syntax very similar to how the `lock` keyword works with all the benefits of async/await.

### Proposal

It is my understanding that it's discouraged/forbidden to reuse reserved keywords for multiple purposes. Given that, I would like to propose the additions that would allow for the following syntax.

```csharp
public class SynchronizedAuthorizationClient {
    private static readonly SemaphoreSlim _semaphore = new(1,1);
    ...

    public async Task<string?> GetAccessToken()
    {
         boundary(_semaphore) {
             await _semaphore.WaitAsync();
         
             var cachedToken = _tokenStorage.GetToken();

            if (cachedToken != null) {
                return cachedToken;
            }

            var token = await _identityProviderClient.GetToken();
            _tokenStorage.SetToken(token);  

            return token;
        }
    }
}
```

Where boundary keyword would be defined as
```csharp
boundary(IReleaseable) {

// Here goes the code that requires synchronization

} // at the end of the boundary, `IDisposable.Dispose` is automatically invoked, on the object returned from `IReleaseable.Wait`
```
and the `IReleaseable` (the name is irrelevant) interface could be defined as
```csharp
interface IReleaseable {
    IDisposable Acquire();
    Task<IAsyncDisposable> AcquireAsync();
    
}
```

Few things to note on the above:
- ideally, `IDisposable` and `IAsyncDisposable` returned from `IReleaseable`, would be replaced by a dedicated interface, `IReleaseableHandle` or similar
- That'd make `IReleaseableHandle` definition to avoid using `Dispose` and `DisposeAsync` methods to avoid confusion/mixing slightly different concepts with one another
- `IReleaseable` interface would then be implemented by the existing synchronization primitives that meet the requirements (`SemaphoreSlim` being one example) or a whole new implementation could be provided, one that is tailored to this type of scenarios 

To summarize, the code that we just saw in the first code excerpt of this section, would translate to the following

```csharp
public class SynchronizedAuthorizationClient {
    private static readonly IReleaseable _sync = new ReleseableImplementation();
    ...

    public async Task<string?> GetAccessToken()
    {
         try {
             var syncHandler = await _sync.AcquireAsync();

             var cachedToken = _tokenStorage.GetToken();

            if (cachedToken != null) {
                return cachedToken;
            }

            var token = await _identityProviderClient.GetToken();
            _tokenStorage.SetToken(token);  

            return token;
        } finally {
            await syncHandler.ReleaseAsync();
        }
    }
}
```
In a synchronous scenario, `AcquireAsync` and `ReleaseAsync` methods would be replaced by `Acquire` and `Release` respectively.

#### Potential additions/amendments
- depending on the implementation, `AcquireAsync` and `ReleaseAsync` (especially the last one) could be removed
- if it made sense to keep the aforementioned methods, an optional parameter could be added to `boundary(IReleaseable)` to force synchronous acquisition and release [`boundary(IReleasable, bool forceSync)`]
- depending on how complex it would be, ideally, explicit usage of the `IReleaseable` implementations would be forbidden. This would force the use of the `boundary` keyword and enforce safety

### Implementation
This is where I lack information currently and would be happy to discuss that further. 

Since the main goal is to improve safety, I believe that even as a mere syntactic sugar, this would be beneficial to C# users who often deal with synchronization and have to deal with cognitive overhead related to releasing synchronization primitives. Just like the `using` keyword does for handling disposables.

Having said that, building it in a way that would not only bring safety but also potential performance improvements would most definitely be welcomed by the community.

I'd be more than happy to author or co-author the implementation should this proposal be taken into consideration.

### [Footnotes]
[1] One could argue that you do not need a semaphore in the presented scenario if you're happy with the consequences of not synchronizing at this level (assuming that the token storage itself is thread-safe). This is beyond the point I'm trying to make here, so I'd appreciate it if people directed their focus to the proposal. I'd be more than happy to replace it with a better example - suggestions are more than welcome!

**Comment by svick** [comment]
Personally, I don't see the need for a new special-purpose syntax here, when you can use the existing general-purpose `using` syntax. The simplest approach would be something like this:

```c#
public async Task<string?> GetAccessToken()
{
    using (await semaphore.AcquireAsync())
    {
        var cachedToken = _tokenStorage.GetToken();

        if (cachedToken != null)
        {
            return cachedToken;
        }

        var token = await _identityProviderClient.GetToken();
        _tokenStorage.SetToken(token);

        return token;
    }
}

static class Extensions
{
    public static async ValueTask<SemaphoreGuard> AcquireAsync(this SemaphoreSlim semaphore)
    {
        await semaphore.WaitAsync();
        
        return new SemaphoreGuard(semaphore);
    }
}

public struct SemaphoreGuard : IDisposable
{
    private SemaphoreSlim semaphore;
    
    internal SemaphoreGuard(SemaphoreSlim semaphore) => this.semaphore = semaphore;
    
    public void Dispose() => semaphore.Release();
}

**Reply by CyrusNajmabadi** [reply]
Note: this pattern is what we use in Roslyn ourselves.

**Reply by 333fred** [reply]
I just added an extension like this in a personal project last week too.

**Reply by HaloFour** [reply]
I'm thinking that these should be considered as standard APIs in the BCL to encourage this pattern.  I think it also lends to the use of `using` for unit-of-work scenarios.

**Reply by pzaj2** [reply]
This is the very same pattern I suggested at the company where I work (hence the references to disposables in my proposal). While it does solve the problem, you need people to remember to use it 😉

Whether it's built as a language feature or added to the standard APIs of the BCL - like @HaloFour suggested - it'd definitely lead to widespread use of this pattern. 

Many features are added, at various levels, just to make solving common problems more convenient and easier. I'd say that as long as a better solution to this somewhat hard-to-debug problem is provided, I'd consider my "mission" successful.

@HaloFour, any chance you could point me in the right direction as to where I could make such a proposal with respect to the BCL? That is if there's a similar process to this one.

**Reply by HaloFour** [reply]
Yes, the [runtime](https://github.com/dotnet/runtime) repository is where runtime and BCL/API requests are handled.
